Add periodic callhome functionality

anjalshireesh opened the pull request:
## Description

Periodically (every 24hrs by default), fetch callhome information and
upload it to SUBNET. As of now the callhome payload contains only
the `admin info` data. It will subsequently be enhanced to send other
useful information.
  
New config keys added under the new `callhome` subsystem:
  
`enable` - Set to `on` for enabling callhome. Default `off`
`frequency` - Interval between callhome cycles. Default `24h`

## Motivation and Context

Send callhome information to SUBNET for better supportability

## How to test this PR?

- For local testing, ensure that the env variable `MINIO_CI_CD` is set to `on`
- Start the `SUBNET` backend locally (it runs on `https://localhost:9000`)
- Start a cluster with `minio` built using this PR
- Enable callhome by running `mc admin config set {alias} callhome enable=on frequency=20s`
- Check that entries are getting created in the SUBNET local db every 20 seconds
`SELECT id, created_at, info from customer.callhome order by id desc LIMIT 10`

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

anjalshireesh committed:
Add periodic callhome functionality

Periodically (every 24hrs by default), fetch callhome information and
upload it to SUBNET.

New config keys under the `callhome` subsystem:

enable - Set to `on` for enabling callhome. Default `off`
frequency - Interval between callhome cycles. Default `24h`

harshavardhana commented on cmd/callhome.go:
```diff
@@ -0,0 +1,133 @@
+	if err != nil {
+		return
+	}
+	defer locker.Unlock(lkctx.cancel)
```
```suggestion
```

anjalshireesh replied:
If I don't unlock, the following scenario doesn't work:

- Callhome enabled (leader acquires lock and starts processing in the loop)
- Callhome disabled (`mc admin config set {alias} callhome enable=off`). The loop breaks and `runCallhome` returns without releasing the lock
- Callhome enabled again (`mc admin config set {alias} callhome enable=on`). The callhome can't start again now because it can't acquire the lock.

harshavardhana replied:
yes but you are doing things wrong here, your locking and unlocking should look like this. 

```
	// Make sure only 1 callhome is running on the cluster.
	locker := objAPI.NewNSLock(minioMetaBucket, "callhome/runCallhome.lock")
	lkctx, err := locker.GetLock(ctx, callhomeLeaderLockTimeout)
	if err != nil {
		return
	}
	ctx = lkctx.Context()
	defer locker.Unlock(lkctx.Cancel)
```

anjalshireesh replied:
Have changed the code as suggested by you

harshavardhana commented on cmd/callhome.go:
```diff
@@ -0,0 +1,133 @@
+	}()
+}
+
+func runCallhome(pctx context.Context, objAPI ObjectLayer) {
```
```suggestion
func runCallhome(ctx context.Context, objAPI ObjectLayer) {
```

anjalshireesh replied:
Done

harshavardhana commented on cmd/callhome.go:
```diff
@@ -0,0 +1,133 @@
+func runCallhome(pctx context.Context, objAPI ObjectLayer) {
+	// Make sure only 1 callhome is running on the cluster.
+	locker := objAPI.NewNSLock(minioMetaBucket, "callhome/runCallhome.lock")
+	lkctx, err := locker.GetLock(pctx, callhomeLeaderLockTimeout)
```
```suggestion
	lkctx, err := locker.GetLock(ctx, callhomeLeaderLockTimeout)
```

anjalshireesh replied:
Done

harshavardhana commented on cmd/callhome.go:
```diff
@@ -0,0 +1,133 @@
+	}
+	defer locker.Unlock(lkctx.cancel)
+
+	ctx := lkctx.Context()
```
```suggestion
	ctx = lkctx.Context()
```

anjalshireesh replied:
Done

harshavardhana commented on cmd/callhome.go:
```diff
@@ -0,0 +1,133 @@
+			// Reset the timer for next cycle.
+			callhomeTimer.Reset(callhomeFreq.Get())
+
+			go performCallhome(ctx)
```
we cannot do an uncontrolled function invocation like this inside a timer, if there is an going callhome in progress it should block and complete gracefully. 

```suggestion
			performCallhome(ctx)
			
			// Reset the timer for next cycle.
			callhomeTimer.Reset(callhomeFreq.Get())
```

anjalshireesh replied:
Done

harshavardhana commented on cmd/subnet-utils.go:
```diff
@@ -0,0 +1,130 @@
+	return &http.Client{
+		Timeout: timeout,
+		Transport: &http.Transport{
+			Proxy: ieproxy.GetProxyFunc(),
```
```suggestion
			Proxy: http.ProxyFromEnvironment,
```

is enough

harshavardhana replied:
Please re-use one of the global transports available, do not need to recreate a new transport every time.

anjalshireesh replied:
Done. Using `globalProxyTransport`

harshavardhana commented on cmd/subnet-utils.go:
```diff
@@ -0,0 +1,130 @@
+			return nil, err
+		}
+		client.Transport.(*http.Transport).Proxy = http.ProxyURL(proxyURL)
+	}
```
```suggestion
	if len(globalSubnetConfig.Proxy) > 0 {
		proxyURL, err := url.Parse(globalSubnetConfig.Proxy)
		if err != nil {
			return nil, err
		}
		client.Transport.(*http.Transport).Proxy = http.ProxyURL(proxyURL)
	}
```

globalSubnetConfig.ProxyURL should already be parsed during MinIO startup and kept in memory, there is no reason to parse this everytime you make a HTTP request.

anjalshireesh replied:
Done

harshavardhana commented on cmd/subnet-utils.go:
```diff
@@ -0,0 +1,130 @@
+		return "", err
+	}
+
+	defer resp.Body.Close()
```
make sure to drain the resp.Body - use xhttp.DrainBody()

anjalshireesh replied:
Done

harshavardhana commented on cmd/subnet-utils.go:
```diff
@@ -0,0 +1,130 @@
+
+func subnetBaseURL() string {
+	if globalIsCICD {
+		return "http://localhost:9000"
```
does this mean that subnet will be running also on port 9000? how can this be tested? - MinIO also uses port 9000

anjalshireesh replied:
Yes, subnet development server runs on 9000. I usually test by running my minio on a different port using `--address :9010`

harshavardhana commented on cmd/subnet-utils.go:
```diff
@@ -0,0 +1,130 @@
+
+func subnetCallhomeURL() string {
+	return subnetBaseURL() + "/api/callhome"
+}
```
if callHomeURL is static why shouldn't this be a constant? - perhaps you can have two constants one for local testing and another for real endpoint.

anjalshireesh replied:
Done

harshavardhana commented on cmd/subnet-utils.go:
```diff
@@ -0,0 +1,130 @@
+
+func sendCallhomeInfo(ch CallhomeInfo) error {
+	if len(globalSubnetConfig.APIKey) == 0 {
+		return errors.New("Cluster is not registered with SUBNET.")
```
perhaps provide a link to "register"

anjalshireesh replied:
As this is not a message shown to the end user, I think it may not be required.

harshavardhana replied:
you are still logging right @anjalshireesh - isn't that useful to show?

anjalshireesh replied:
Have appended the error message with `Please register by running 'mc support register ALIAS'`

harshavardhana commented on cmd/subnet-utils.go:
```diff
@@ -0,0 +1,130 @@
+
+func subnetAuthHeaders(authToken string) map[string]string {
+	return map[string]string{"Authorization": authToken}
+}
```
this looks a bit clunky, why do you need a map to set key value to another map? why not pass the authToken to subnetPostReq() and then directly add it to `req.Headers.Set()`

anjalshireesh replied:
Thanks, done.

harshavardhana commented on internal/config/callhome/callhome.go:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	cfg.Enable = env.Get(config.EnvMinIOCallhomeEnable,
+		kvs.GetWithDefault(Enable, DefaultKVS)) == "on"
```
```suggestion
		kvs.GetWithDefault(Enable, DefaultKVS)) == config.EnableOn
```

anjalshireesh replied:
Done

harshavardhana commented on cmd/callhome.go:
```diff
@@ -0,0 +1,133 @@
+
+func updateCallhomeParams(ctx context.Context, objAPI ObjectLayer) {
+	alreadyEnabled := enableCallhome
+	enableCallhome = globalCallhomeConfig.Enable
```
this is racy, why do you need to replace a global value right like this when you already have this as part of another global variable?

harshavardhana replied:
if you want to do it safely, keep a mutex.

anjalshireesh replied:
> this is racy

This is executed whenever there is a change in the callhome config, which is done very rarely and triggered manually. So raciness is not a concern I think.

However I'll take your point and have modified the code to update and access `enableCallhome` safely using a mutex.

> why do you need to replace a global value right like this when you already have this as part of another global variable?

This helps in detecting when callhome was disabled earlier and has been enabled now, allowing us to initialize the callhome process.

klauspost replied:
> done very rarely [...] So raciness is not a concern I think.

Raciness is always a concern. There are no "harmless" races.

anjalshireesh replied:
@klauspost point taken. Have already addressed the concern. PTAL.

anjalshireesh committed:
Improvements based on review comments

- Update `enableCallhome` safely
- Rename pctx to ctx
- Block during execution of callhome
- Store parsed proxy URL in global subnet config
- Store callhome URL(s) in constants
- Use existing global transport
- Pass auth token to subnetPostReq
- Use `config.EnableOn` instead of `"on"`

klauspost commented on cmd/callhome.go:
```diff
@@ -0,0 +1,154 @@
+	s.Lock()
+	defer s.Unlock()
+	return s.b
+}
```
We might as well make it lockless:

```suggestion
// safeBool contains an atomic bool value.
// Zero value is "false", use safeBoolTrue to initialize a true value.
type safeBool int32

const (
	// Use for initializing only, not for comparisons
	safeBoolFalse safeBool = 0
	safeBoolTrue  safeBool = 1
)

const (
	// Use for initializing only, not for comparisons
	safeBoolFalse safeBool = 0
	safeBoolTrue  safeBool = 1
)

// Update will update the bool.
// Order of concurrent updates is not guaranteed
func (s *safeBool) Update(b bool) {
	if b {
		atomic.StoreInt32((*int32)(s), 1)
	} else {
		atomic.StoreInt32((*int32)(s), 0)
	}
}

// Get returns the bool value.
func (s *safeBool) Get() bool {
	return atomic.LoadInt32((*int32)(s)) == 1
}
```

You are welcome to do the same to `safeDuration` (hint: time.Duration is an int64)

anjalshireesh replied:
Thanks. Have used this, and tried to do the same for `safeDuration` as well. PTAL.

klauspost commented on internal/config/callhome/help.go:
```diff
@@ -0,0 +1,42 @@
+	HelpCallhome = config.HelpKVS{
+		config.HelpKV{
+			Key:         Enable,
+			Type:        "boolean",
```
```suggestion
			Type:        "on|off",
```
(as used elsewhere)

anjalshireesh replied:
Thanks, done.

anjalshireesh committed:
Use atomic package instead of lock

klauspost reviewed (approved):
LGTM

harshavardhana commented on cmd/data-scanner.go:
```diff
@@ -88,21 +86,15 @@ func initDataScanner(ctx context.Context, objAPI ObjectLayer) {
-	defer s.Unlock()
-	return s.t
+	return time.Duration(atomic.LoadInt64((*int64)(s)))
 }
```
Since we already use `go.uber.org/atomic`  we should just use `uatomic.NewDuration()` @anjalshireesh

anjalshireesh replied:
Done

harshavardhana commented on cmd/subnet-utils.go:
```diff
@@ -0,0 +1,105 @@
+	ct := r.Header.Get("Content-Type")
+	if len(ct) == 0 {
+		r.Header.Add("Content-Type", "application/json")
+	}
```
why do we need this flexibility? we know that its always going to be `application/json` right? @anjalshireesh is there a possibility of sending any other content-type?

anjalshireesh replied:
Not required. Removed.

harshavardhana commented on internal/config/subnet/api-key.go:
```diff
@@ -49,7 +49,7 @@ type Config struct {
 
 	// The HTTP(S) proxy URL to use for connecting to SUBNET
-	Proxy string `json:"proxy"`
+	ProxyURL *url.URL `json:"proxy_url"`
```
you can our project here *xnet.URL - does a more stricter parsing. and also parse the URL as `xnet.ParseHTTP()`

anjalshireesh replied:
Done

harshavardhana commented on cmd/callhome.go:
```diff
@@ -0,0 +1,159 @@
+// Get returns the bool value.
+func (s *safeBool) Get() bool {
+	return atomic.LoadInt32((*int32)(s)) == 1
+}
```
SafeBool not needed use `uatomic.NewBool()`

anjalshireesh replied:
Done

anjalshireesh committed:
Use uber atomic package

harshavardhana commented on cmd/callhome.go:
```diff
@@ -0,0 +1,136 @@
+	}
+
+	ctx = lkctx.Context()
+	defer locker.Unlock(lkctx.cancel)
```
```suggestion
	defer locker.Unlock(lkctx.Cancel)
```

anjalshireesh committed:
Use `Cancel` instead of `cancel`

Co-authored-by: Harshavardhana <harsha@minio.io>

donatello committed:
Merge branch 'master' into callhome
